### DISCUSSION TITLE
TestObserver implementation question

### DISCUSSION THREAD

**Discussion by john-flanagan** [pointfreeco/swift-dependencies#66] [General]
I'm trying to better understand test dependency caching. If I'm understanding correctly this code clears out the dependency cache between each test case:
https://github.com/pointfreeco/swift-dependencies/blob/51721ee26506cb10ebe76d21de60ebae0501455a/Sources/Dependencies/DependencyValues.swift#L355-L397

I'm just curious why the implementation is using `objc_getProtocol`, `NSClassFromString`, etc when Objective-C is available rather than always using `XCTestObservationCenter.shared.addTestObserver(TestObserver())`.

**Comment by MinLofftP** [comment]
As for why the implementation is using objc_getProtocol and NSClassFromString when Objective-C is available, I'm not entirely sure. However, my best guess is that the code is trying to provide a fallback mechanism for when XCTestObservationCenter is not available, as XCTestObservationCenter.shared is only available on iOS 9.0 and later. By using these Objective-C methods, the code is able to work on earlier versions of iOS that don't have XCTestObservationCenter.

I hope that helps! Let me know if you have any other questions or if there's anything else I can assist you with.

**Comment by john-flanagan** [comment]
I don't know how I missed the comment directly above the code I highlighted 🤦 
```swift
// NB: We cannot statically link/load XCTest on Apple platforms, so we dynamically load things
//     instead and we limit this to debug builds to avoid App Store binary rejection.
```
